### ISSUE TITLE
Rust: Work with stable Rust

### ISSUE THREAD

**Issue by chrysn** [RIOT-OS/RIOT#17526]
### Description

With the `asm!` macro stabilized for Rust 1.59 (provided there are no reverts until or after the beta snapshot), using a stable Rust compiler should be possible. (So far, features were selected by the dozen because why not).

### Testing procedure

Until there is a stable Rust that has all features, you can pick the ones you're willing to tolerate still using:

```
$ RUSTFLAGS="-Zallow-features=never_type" make all -C examples/riot-hello-world BOARD=particle-xenon
```

In my cortex-m3 tests on the relevant branches (see below), the `never_type` feature is the only one that was not worked around yet (not because it doesn't work but because many lines need changing).

### Useful links

Most of the current status is described at https://forum.riot-os.org/t/using-rust-with-riot-os/208/26 (thanks to geonnave and @mcr for pushing here).

Both riot-sys and riot-wrappers have `stable-rust` branches that are currently a mix of clipping out things that could long have worked on stable, and doing actual changes to become runnable on stable.

**Comment by chrysn** [comment]
When this is closed, https://arewertosyet.com/ will need an update.

**Comment by chrysn** [comment]
Wow, 1.59 is already in beta.

I've done one very crude test with the remaining `never_type` feature (essentially commenting out the GPIO and LED modules as well as some termination stuff), and behold:

```
$ CARGO_CHANNEL=beta make all -C examples/rust-hello-world BOARD=microbit-v2 all flash term
[... including a press to reset to capture boot output ...]
2022-01-17 18:05:03,717 # main(): This is RIOT! (Version: 2022.04-devel-46-gd788bd)
2022-01-17 18:05:03,718 # Hello Rust!
```

**Comment by chrysn** [comment]
This has progressed now with the release of riot-sys 0.7.1 and riot-wrappers 0.7.16. (No lockstepping by the way, just both happen to be around 0.7 these days).

Current test procedure is:
* Patch any example's dependency line to say `riot-wrappers = { version = "0.7", default-features = false, features = ["set_panic_handler", "panic_handler_format" ] }`. (That disables the cstr_nightly feature that is recently default for compatibility reason; that workaround will go away with 0.8 which will reap a few deprecations and this transition).
* Run `cargo update`
* Run `CARGO_CHANNEL=beta make all BOARD=microbit-v2`

(Note that this only works on the hello-world and minimal test, and only on ARM boards: The riot-gcoap example use an external crate (coap-message) that uses some unstable features, the transpilate of RISC-V needs some unsafe stuff, and building for native needs an unstable language item in riot-wrappers -- but still, for some cases (and I dare say the more common ones), things are progressing.

I'm holding off on actually updating the examples until at least a riot-wrappers 0.8 is released that makes the patch unnecessary (the examples would then depend on 0.8, or `>= 0.7 < 0.9` if there's any value to it), and/or until 1.59 becomes stable, at which point it makes sense to update the "Using Rust" documentation.

**Comment by chrysn** [comment]
Actually, even native can be built on stable if `panic = "abort"` is set in the used profile (which doesn't make a lot of difference given the panic handler used diverges anyway). Won't put much effort in exposing this unless there is a need for having this particular platform available on stable as well.

**Comment by chrysn** [comment]
What remains after #17704 is that riot-wrappers still has cstr_nightly as a default feature, that needs to be disabled.

Not sure if it's worth adjusting the examples, or whether riot-wrappers shouldn't just do a 0.8 bump that just reaps all deprecatings including the cstr_nightly default. (In that bump it could also remove the panic handler setting, which would be really better done explicitly / opt-in to resolve <https://gitlab.com/etonomy/riot-wrappers/-/merge_requests/6>).

.... and to fully close this I'd also like to have it checked in CI (and I'm not sure how happy we are with yet another compiler inside the image, gotta do numbers).

Anyway, current patch for the rust-hello-world example to work on stable:

```patch
diff --git a/examples/rust-hello-world/Cargo.toml b/examples/rust-hello-world/Cargo.toml
index 443cf57cf2..47d8ebaf6b 100644
--- a/examples/rust-hello-world/Cargo.toml
+++ b/examples/rust-hello-world/Cargo.toml
@@ -11,2 +11,2 @@ crate-type = ["staticlib"]
 [dependencies]
-riot-wrappers = "0.7"
+riot-wrappers = { version = "0.7", default-features = false, features = [ "set_panic_handler", "panic_handler_format" ] }
```

```
$ make all term CARGO_CHANNEL=stable BOARD=microbit-v2 all flash term
```

**Comment by chrysn** [comment]
Things are converging for even more stability: Outside of per-platform (nightly panic handler, riscv specials) and optional dependencies (coap-message has nightly features in its API), the two remaining widely used (SAUL, shell) of riot-wrappers are const_fn_fn_ptr_basics and const_fn_trait_bound, and these are part of a [stabilization round](https://github.com/rust-lang/rust/pull/93827) that is on the road to be merged and would be part of Rust 1.61.
